File: src/main/java/me/testaccount666/serversystem/updates/UpdateCheckerType.java
package me.testaccount666.serversystem.updates;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.Arrays;
import java.util.Optional;
import java.util.function.Supplier;

@RequiredArgsConstructor
@Getter
public enum UpdateCheckerType {
    DISABLED(DisabledUpdateChecker::new),
    HANGAR(HangarUpdateChecker::new),
    MAIN(MainUpdateChecker::new);

    private final Supplier<AbstractUpdateChecker> _factory;

    public static Optional<UpdateCheckerType> of(String name) {
        return Arrays.stream(values()).filter(type -> type.name().equalsIgnoreCase(name)).findFirst();
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/updates/UpdateCheckerType.kt
package me.testaccount666.serversystem.updates

enum class UpdateCheckerType(val new: () -> AbstractUpdateChecker) {
    DISABLED({ DisabledUpdateChecker() }),
    HANGAR({ HangarUpdateChecker() }),
    MAIN({ MainUpdateChecker() });

    companion object {
        fun of(name: String) = entries.firstOrNull { it.name.equals(name, ignoreCase = true) }
    }
}